File: client/lib/cache.ts
// Shared in-memory LRU cache for LLM results

const MAX_CACHE = 200;
const llmCache = new Map<string, unknown>();

export function cacheKey(...parts: string[]): string {
  const str = parts.join("\x00");
  let h = 5381;
  for (let i = 0; i < str.length; i++) {
    h = ((h << 5) + h) ^ str.charCodeAt(i);
    h = h >>> 0;
  }
  return h.toString(36);
}

export function cacheGet<T>(key: string): T | undefined {
  return llmCache.get(key) as T | undefined;
}

export function cacheSet(key: string, value: unknown) {
  if (llmCache.size >= MAX_CACHE) {
    const firstKey = llmCache.keys().next().value;
    if (firstKey !== undefined) llmCache.delete(firstKey);
  }
  llmCache.set(key, value);
}

File: PromptWars_Aetherion_project/PromptWars_Aetherion/lib/cache.py
from collections import OrderedDict
from typing import Any, Optional

MAX_CACHE = 200
_llm_cache: OrderedDict[str, Any] = OrderedDict()


def cache_key(*parts: str) -> str:
    text = "\x00".join(parts)
    h = 5381
    for ch in text:
        h = ((h << 5) + h) ^ ord(ch)
        h = h & 0xFFFFFFFF
    return format(h, "x")


def cache_get(key: str) -> Optional[Any]:
    return _llm_cache.get(key)


def cache_set(key: str, value: Any) -> None:
    if len(_llm_cache) >= MAX_CACHE:
        _llm_cache.popitem(last=False)
    _llm_cache[key] = value
